### ISSUE TITLE
[3.0 & 2.1]: Database Error: syntax error, unexpected '+', expecting FTS_TERM or FTS_NUMB or '*'

### ISSUE THREAD

**Issue by sbulen** [SimpleMachines/SMF#8350]
### Basic Information

I have started getting errors on my 2.1 forum, which I had recently cut over to using fulltext search.  (Previously I had used a custom index, I was trying to see if a fulltext search would resolve other search issues, see #7870...)  

I have not been able to reproduce.  I am assuming both 2.1 & 3.0 are affected since the codebases are similar.  

I was hoping someone familiar with the fulltext syntax could look into this.

The errors are:
Database Error: syntax error, unexpected '+', expecting FTS_TERM or FTS_NUMB or '*'
Database Error: syntax error, unexpected ')', expecting $end

Both errors are on /Sources/SearchAPI-Fulltext.php Line 297
Both errors are on `https://yadayada/index.php?action=search2`

### Steps to reproduce

1.  I do not know how to reproduce at this time

### Version/Git revision

2.1.4

### Database Engine

MySQL

### Database Version

8.0.36

### PHP Version

8.2.24

### Logs

```shell
Errors have the same backtrace:
    #0: smf_db_error()
    Called from /yada/yada/yada/smf/Sources/Subs-Db-mysql.php on line 494
    #1: smf_db_query()
    Called from /yada/yada/yada/smf/Sources/SearchAPI-Fulltext.php on line 297
    #2: indexedWordQuery()
    Called from /yada/yada/yada/smf/Sources/Search.php on line 1588
    #3: PlushSearch2()
    Called from /yada/yada/yada/smf/index.php on line 224
```

**Comment by jdarwood007** [comment]
@sbulen can you capture the query being generated?  I suspect it is not encoding something properly when building the query.
Maybe related to MB4 as well.

**Comment by sbulen** [comment]
No, I haven't caught one yet.  Still getting them.  Forum is standard smf utf8mb3.

They are pretty infrequent - only a couple a month.  So it's some weird condition.

**Comment by jdarwood007** [comment]
I meant that we are seeing a MB4 string come in, and its not being encoded properly.
I'm not a fulltext expert either, but it looks like what is failing is the insert select statement that dumps the results into a temp table for working through.  Which means we are more likely failing to handle something on that query rather than anything else.

**Comment by sbulen** [comment]
I've done a lot of testing with mb4 characters to attempt to reproduce & had no issues...

My suspicion is it's more a matter of an unescaped character borking the search.  But I haven't been able to reproduce manually.

**Comment by jdarwood007** [comment]
@sbulen, I think I should close this up for now.  You can always reopen it later or create a new one.

1. It's not reproducible
2. We don't know what query it was running in full
3. We don't know the search terms, and parameters being issued

I do wonder if it's related to #8312 (but that one uses the custom search index) in the fact that we do split up search terms and then try to parse/handle them, and possibly we are silently triggering the same truncation error, resulting in some bad UTF8 being run in the query.

**Comment by sbulen** [comment]
Agreed.  I was kinda hoping someone more familiar with that syntax would have a clue & point us in the right direction.

**Comment by sbulen** [comment]
Timing...  I had a couple of these happen recently and the queries were logged this time...  At least part of the problem appears to be trailing syntax elements, e.g:

`larry)` will get the "syntax error, unexpected ')', expecting $end" error
`larry+` will get the "syntax error, unexpected $end" error

Some online searches show that the other error `syntax error, unexpected '+', expecting FTS_TERM or FTS_NUMB or '*'` was actually a MySQL bug when searching for * on certain versions of MySQL under InnoDB.  I cannot reproduce, & my DB has been patched since then, so for the moment, I'll assume that was due to that MySQL bug.

**Comment by sbulen** [comment]
Not sure if you want to reopen this or not...

**Comment by jdarwood007** [comment]
I will reopen pending trying to reproduce the updated errors.

**Comment by jdarwood007** [comment]
Going to relose this with some notes.

The bugged version of MySQL, from what I can find, was 5.6/5.7, and you reported using 8.0, so it shouldn't be affected, but maybe poor docs.
Additionally, I tested using multiple variances of `test*`, `test+`, `"test"*` in both the search params, as a username, and changed all the check boxes.  I also tested against both fulltext and custom.  Neither one could reproduce it.

For now, I think it's best to close again, pending more information that helps reproduce the issue, so we can understand what needs to be fixed.
